advanced-achievements-plugin/src/main/java/com/hm/achievement/db/CacheManager.java

```diff
@@ -67,6 +68,23 @@ public CacheManager(AdvancedAchievements advancedAchievements, @Named("main") Co
 			multipleAchievementsToPlayerStatistics.put(multipleAchievement, new ConcurrentHashMap<>());
 		}
 		totalPlayerAchievementsCache = new ConcurrentHashMap<>();
+
+	}
+
+	public Map<UUID, Set<String>> getReceivedAchievementsCache() {
+		return receivedAchievementsCache;
+	}
+
+	public Set<Achievement> getCache() {
+		return cache;
+	}
+
+	public void populateCache(UUID uuid) {
+		CompletableFuture.runAsync(() -> {
```

Comment:
How would I query that? going through all of the achievements and checking for completion?

Reply:
Something like this would work I think:
````
			Set<String> playerNotReceivedAchievementNamesList = namesToDisplayNames.keySet().stream()
					.filter(name -> !playerAchievementNamesList.contains(name))
					.collect(Collectors.toSet());
			notReceivedAchievementsCache.put(uuid, playerNotReceivedAchievementNamesList);
````

You simply need to inject `@Named("ntd") Map<String, String> namesToDisplayNames` in the class.

Another way of looking at the problem is now that the cache is populated eagerly, there's in no point in keeping a separate cache structure `notReceivedAchievementsCache`: no received will simply be equivalent to `!receivedAchievementsCache.containsKey(uuid)`. That's a bit more work to do in this class, but it will simplify things a fair bit, up to you to decide. I'm also happy to do the cleanup in a follow up commit, in which case you can drop this review comment entirely. 😉